firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteSchema.java

```diff
@@ -364,10 +364,11 @@ private void createFieldIndex() {
           db.execSQL(
               "CREATE TABLE index_entries ("
                   + "index_id INTEGER, "
-                  + "index_value BLOB, " // field value pairs
+                  + "array_value BLOB, " // index values for ArrayContains/ArrayContainsAny
+                  + "directional_value BLOB, " // index values for equality and inequalities
                   + "uid TEXT, " // user id or null if there are no pending mutations
                   + "document_name TEXT, "
-                  + "PRIMARY KEY (index_id, index_value, uid, document_name))");
+                  + "PRIMARY KEY (index_id, array_value, directional_value, uid, document_name))");
```

Comment:
Only applying the query limit locally can be really inefficient since we would always read all matching values even if the limit is low. This then imposes ordering, which unfortunately makes it necessary to split up array values from the other field values (as explained in the PR description). 

We can actually use the array_value here even if it is not set by simply enforcing that it is null. We already don't use filters that have an array_value for non ArrayContains/ArrayContains any queries.

```
EXPLAIN QUERY PLAN select * from index_entries_test where (index_id=1 and array_value IS NULL AND directional_value == 'x') order by directional_value limit 2;
QUERY PLAN
`--SEARCH TABLE index_entries_test USING COVERING INDEX sqlite_autoindex_index_entries_test_1 (index_id=? AND array_value=? AND directional_value=?)
```

Reply:
Right, always having a filter on `array_value` should fix this. 

Please add a comment to document this in SQLiteIndexManager when the query is built.

Reply:
Also, `limit` is probably still done in memory, by SQLite, but I think we can live with that.

Reply:
Will merge this PR as is and add the comment in the follow up PR (the one that fixes IN queries).